### PR TITLE
[WIP] Parallelize the compiler via two-pass compilation

### PR DESCRIPTION
**Pull request by smarter**
How do you parallelize a compiler? A first approach might be: given a list of source files, split it into N groups then compile it using N compiler instances. If each file can be compiled independently this works, but usually files will refer to symbols defined in other files, so we need to know in advance about these symbols and their types. If we had previously compiled these files, this is easy: we can just read the compiler output (.class and .tasty files) to find every symbol and its type, this is how incremental compilation already works. Of course, we can't assume that the files we're trying to compile have already been compiled, but we don't need the full compiler output: we're only intested in the name and type of symbols. From this, we can sketch a simple way to parallelize the compiler:

1. Run a first pass on the input files that computes the type of every (non-private) definition, store the result in memory. This needs to be significantly faster than running the full compiler for this approach to be viable.
2. Run a second pass that splits the input files into N groups and compile each group independently, using the information from the first-pass to handle inter-group dependencies.

This is what this PR implements using the flag `-parallelism N`. The first pass works by running the compiler in a special mode where:
- The body of a definition (def, val or var) is not typechecked and its body is replaced by `???` (unless its result type needs to be inferred, or it's a special case, see Typer#canDropBody).
- Statements in a class body are dropped.
- .tasty files are emitted for Java source files too.
- Compilation is stopped after the Pickler phase (which will emit both .tasty files as well as empty .class files).

When compiling Dotty itself, the first pass compile time is about 20% of a regular non-parallel compilation (it'd probably be quite a bit less if we added explicit result types to every definition, I haven't tried that yet).

The second pass just runs N instances of the regular compiler, then combines the results.

The implementation complexity of this approach is very low (we only need to add a few lines of the code to the typechecker), and already gives interesting results. I haven't done any rigorous benchmarking yet, but here's what I get on my laptop (quad core, with hyper-threading enabled) when compiling Dotty itself:

|                               | -parallelism 1 (normal single-threaded compilation) | -parallelism 2 | -parallelism 4 | -parallelism 8 |
|-------------------------------|----------------|----------------|----------------|----------------|
| Compile time                  | 18.7 seconds   | 12.8 seconds   | 10.7 seconds   | 9.4 seconds    |
| Speed-up compared to baseline | 1x             | 1.46x          | 1.74x          | 1.99x          |

I suspect we could improve this significantly by implementing a work-stealing algorithm to decide which file will be compiled by which thread instead of simply dividing the list of files into N groups, since threads may sit idle if some groups end up being faster to compiles than other. An intermediate solution would be to get each thread to compile approximately the same number of lines of code instead of the same number of files.

An interesting property of the two-pass approach is that it can be combined with other parallelization techniques if they bear fruits:
- If @gkossakowski [Kentucky Mule](https://github.com/gkossakowski/kentuckymule) or @xeno-by [rsc](https://github.com/twitter/rsc) manage to implement a significantly faster typechecker, we could speed-up our first pass.
- If we figure out a good way to share some of the internal data structures used by the compiler between threads, like what @DarkDimius experimented with in https://github.com/lampepfl/dotty/issues/991, we could speed-up (and reduce the memory requirements) of our second pass.

Note that this PR is based on https://github.com/lampepfl/dotty/pull/4467, so the first 5 commits can be skipped. Note also that this is still a work in progress, but it should already work well enough for people to experiment with it. Enjoy!

### CODE REVIEW

**Comment by smarter** [comment]
It's also worth noting that in a project with two subprojects A and B, where B depends on A, the output of the first-pass compilation for A could be used as input for B to start its compilation earlier instead of waiting for A to finish compiling, @jvican has been experimenting with something like this in the context of scalac.

**Comment by jvican** [comment]
Yes, it will be available in Bloop 1.1.0, together with implementation notes and detailed explanation of how everything works.

**Comment by xeno-by** [comment]
Super exciting results! I like how you've been able to use compilation from Tasty to leverage the signatures produced in the first phase.

Here's our take on this, based on a similar architecture but compatible with Scalac: https://github.com/twitter/rsc/pull/85. We have implemented support for a subset of Scala to produce signatures for automatically rewritten core of [Twitter Util](https://github.com/twitter/util), and we will be working on adding support for more Scala features according to [the roadmap](https://github.com/twitter/rsc/blob/master/docs/roadmap.md).

**Comment by smarter** [comment]
> Here's our take on this, based on a similar architecture but compatible with Scalac

Interesting! I think that there's one significant difference with the approach I'm taking here though, in https://github.com/twitter/rsc/blob/master/docs/language.md you state:
> The biggest limitation is the lack of type inference for result types of vals, vars and defs, as well as type arguments for super constructor calls. We don't plan to address this limitation in the near future. Instead, we rely on Scalafix to automatically rewrite Scala sources to be compatible with Rsc.

By contrast, the parallelism that this PR enables works fine with code bases that do not have explicit result types everywhere, the first pass will just be slower the more it needs to do type inference. As an example, I've [tweaked the core of Twitter Util to make it compile with Dotty](https://github.com/smarter/util/commits/dotty) (I should add it to the [dotty-community-build](https://github.com/lampepfl/dotty-community-build) and report the bugs I worked around too). On my laptop, hot compilation with `-parallel 1` takes approximately 5 seconds and hot compilation with `-parallel 8` takes approximately 3 seconds (I'm running this through sbt which only reports crude time estimates).

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala [hunk 1]
```diff
@@ -46,6 +46,7 @@ class ScalaSettings extends Settings.SettingGroup {
   val rewrite = OptionSetting[Rewrites]("-rewrite", "When used in conjunction with -language:Scala2 rewrites sources to migrate to new syntax")
   val silentWarnings = BooleanSetting("-nowarn", "Silence all warnings.")
   val fromTasty = BooleanSetting("-from-tasty", "Compile classes from tasty in classpath. The arguments are used as class names.")
+  val parallelism = IntSetting("-parallelism", "Number of parallel threads, 0 to use all cores.", 0)
```
[the remark this answers is not in the document]
Comment (smarter) [reply compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: When parallelism is set to 0, we create one thread per "core" (for some definition of core) on your computer (to be more precise, we create`runtime.getRuntime().availableProcessors()` threads).

**Comment by gkossakowski** [comment]
This is really cool! I agree that dividing the typechecking into symbol table typechecking (signatures) and method body typechecking is the way to go. The key variable in this approach is how quickly one can compute the whole symbol table. This is the question I was researching in Kentucky Mule and I became convinced symbol table can be computed really quickly with careful effort. The symbol table calculation itself can be parallelized which came as a surprise to me. I gave a talk on this subject at SF Scala meetup two months ago.

@smarter out of curiosity, do you plan to work more on parallelization of dotty?

PS. I'm in Basel for the summer and happy to chat about this subject.

**Comment by smarter** [comment]
> out of curiosity, do you plan to work more on parallelization of dotty?

Yes, though I have other higher priorities so not sure how much time I'll spend on it.

> PS. I'm in Basel for the summer and happy to chat about this subject.

You should come to EPFL to give a talk :)

**Comment by smarter** [comment]
> out of curiosity, do you plan to work more on parallelization of dotty?

Yes, though I have other higher priorities so not sure how much time I'll spend on it.

> PS. I'm in Basel for the summer and happy to chat about this subject.

You should come to EPFL to give a talk :)

**Comment by odersky** [comment]
@smarter This has been inactive for 6 months. Should we keep it open?

**Comment by smarter** [comment]
Yes, I'm still on it, I thinking keeping PRs marked WIP open is OK.

**Comment by anatoliykmetyuk** [comment]
There was no activity on this one for a long while, so let's close it.

**Pull request closed without merging**